src/output/plugins/OSXOutputPlugin.cxx

```diff
@@ -220,43 +221,116 @@ osx_output_set_device(OSXOutput *oo, Error &error)
+	sake of correctness this callback allows for de-interleaving
+	anyway, and calculates the expected output layout by examining
+	the output buffers.
+	
+	The input buffer is a DynamicFifoBuffer. When a
+	DynamicFifoBuffer contains less data than we want to read from
+	it there is no point in doing a second Read(). So in the case
+	of insufficient audio data in the input buffer, this render
+	callback will emit silence into the output buffers (memset
+	zero).
+*/
+
 static OSStatus
 osx_render(void *vdata,
 	   gcc_unused AudioUnitRenderActionFlags *io_action_flags,
 	   gcc_unused const AudioTimeStamp *in_timestamp,
 	   gcc_unused UInt32 in_bus_number,
-	   gcc_unused UInt32 in_number_frames,
+	   UInt32 in_number_frames,
 	   AudioBufferList *buffer_list)
 {
+	AudioBuffer *output_buffer = nullptr;
+	size_t output_buffer_frame_size, dest;
+
 	OSXOutput *od = (OSXOutput *) vdata;
-	AudioBuffer *buffer = &buffer_list->mBuffers[0];
-	size_t buffer_size = buffer->mDataByteSize;
+	DynamicFifoBuffer<uint8_t> *input_buffer = od->buffer;
+	assert(input_buffer != nullptr);
+
+	/*
+		By convention when interfacing with audio hardware in CoreAudio,
+		in_bus_number equals 0 for output and 1 for input. Because this is an
+		audio output plugin our in_bus_number should always be 0.
+	*/
+	assert(in_bus_number == 0);
+
+	unsigned int input_channel_count = od->asbd.mChannelsPerFrame;
+	unsigned int output_channel_count = 0;
+	for (unsigned int i = 0 ; i < buffer_list->mNumberBuffers; ++i) {
+		output_buffer = &buffer_list->mBuffers[i];
+		assert(output_buffer->mData != nullptr);
+		output_channel_count += output_buffer->mNumberChannels;
+	}
+	assert(output_channel_count == input_channel_count);
 
-	assert(od->buffer != nullptr);
+	size_t input_buffer_frame_size = od->asbd.mBytesPerFrame;
+	size_t sample_size = input_buffer_frame_size / input_channel_count;
 
+	// Acquire mutex when accessing input_buffer
 	od->mutex.lock();
 
-	auto src = od->buffer->Read();
-	if (!src.IsEmpty()) {
-		if (src.size > buffer_size)
-			src.size = buffer_size;
-
-		memcpy(buffer->mData, src.data, src.size);
-		od->buffer->Consume(src.size);
+	auto src = input_buffer->Read();
+
+	UInt32 available_frames = src.size / input_buffer_frame_size;
+	// Never write more frames than we were asked
+	if (available_frames > in_number_frames)
+		available_frames = in_number_frames;
+
+	/*
+		To de-interleave the data in the input buffer so that it fits in
+		the output buffers we divide the input buffer frames into 'sub frames'
+		that fit into the output buffers.
+	*/
+	size_t sub_frame_offset = 0;
+	for (unsigned int i = 0 ; i < buffer_list->mNumberBuffers; ++i) {
+		output_buffer = &buffer_list->mBuffers[i];
+		output_buffer_frame_size = output_buffer->mNumberChannels * sample_size;
+		for (UInt32 current_frame = 0; current_frame < available_frames; ++current_frame) {
+				dest = (size_t) output_buffer->mData + current_frame * output_buffer_frame_size;
+				memcpy(
```

Comment:
If we are asked to fill 1 output buffer with 512 frames (these are numbers I observed in practice) we will call memcpy 512 times.